### ISSUE TITLE
Finetune Contextual shifted CLIP 

### ISSUE THREAD

**Issue by Kimsure** [workforai/SCAN#12]
It is located in SCAN/open_clip_training/src/open_clip/transformer.py at approximately L578.

_Originally posted by @yongliu20 in https://github.com/yongliu20/SCAN/issues/6#issuecomment-2255277993_
            
I didn't find where the CS module is used in class SCAN(Mask2Former), the only one about this is clip_adapter, but it is for frozen clip encoder(self.clip_adapter.original_clip). Could you give some instructions about where the CS is used?

**Comment by SuleBai** [comment]
Hi, CS module is used in the `self.clip_adapter.clip_model`  forward process.

The code [here](https://github.com/yongliu20/SCAN/blob/097fbbdaf7789ac6d2da44533e95521e73fd7c4f/scan/modeling/clip_adapter/adapter.py#L146) is the forward function, and it calls the CS stategy code [here](https://github.com/yongliu20/SCAN/blob/097fbbdaf7789ac6d2da44533e95521e73fd7c4f/open_clip_training/src/open_clip/transformer.py#L578).

Hope this would help you.

**Comment by Kimsure** [comment]
Thanks for your reply, Sry to bother but I have a few more questions.

I understood the codes you post. However in `ovseg_model.py`, the class `SCAN` only use function of `clip_adapter` including `self.clip_adapter.original_clip`, `self.clip_adapter.get_text_features`, `self.clip_adapter.get_sim_logits` and `self.clip_adapter.normalize_feature`. I didn't find where the forward function of `clip_adapter` is used, and I didn't find where the mask proposals from mask2former is used.

Hope for your instructions.

**Comment by SuleBai** [comment]
Hi, `SCAN` use the `clip_adapter` in the inference branch, code is [here](https://github.com/yongliu20/SCAN/blob/097fbbdaf7789ac6d2da44533e95521e73fd7c4f/scan/ovseg_model.py#L302).

And in the `semantic_inference` function, the forward function is called [here](https://github.com/yongliu20/SCAN/blob/097fbbdaf7789ac6d2da44533e95521e73fd7c4f/scan/ovseg_model.py#L327).

Hope this would help you.

**Comment by Kimsure** [comment]
Thanks for your patient explanations that help me a lot. It’s an impressive and insightful work.

So, CS module is applied in the inference but not training stage, `learnable` means the masked patches are partly replaced CS. CLIP encoder ?

**Comment by SuleBai** [comment]
Yes,  the CS module is applied in the inference but not training stage.

As for the `learnble`, you can checkout the instructions in `README`. We follow the previous work `ovseg` to finetune the CLIP encoder.

```
(Optional) If you want to finetune the CLIP model from scratch, please follow [ov-seg](https://github.com/facebookresearch/ov-seg) to prepare the corresponding data. The finetued command is:

cd open_clip_training
cd src
bash scripts/finetune_VitL_with_mask.sh
```

**Comment by Kimsure** [comment]
Understood. Thanks again for your useful reply.
